### ISSUE TITLE
Unable to connect to oauth/gitlab/login

### ISSUE THREAD

**Issue by dn0sar** [mattermost/desktop#2403]
I confirm (by marking "x" in the [ ] below: [x]):

- [x] This is not a troubleshooting question. [Troubleshooting questions go here: https://docs.mattermost.com/install/troubleshooting.html](https://docs.mattermost.com/install/troubleshooting.html).
- [x] This doesn't reproduce on web browsers (such as in Chrome). If it does, [issue reports go to the Mattermost Server repository](https://github.com/mattermost/platform/issues).
- [x] I have read [contributing guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md).

---

**Summary**

My server implements gitlab authentication. When connecting from a web browser everything works as expected. However, on the Desktop app, if I click on the gitlab link on the authentication page, the app does not redirect to the gitlab website. I tried downgrading to the previous version (Version 5.1.1 (5.1.1.20701)) and this issue is not there.

**Environment**
- Operating System: MacOS 13.0
- Mattermost Desktop App version: 5.2.1
- Mattermost Server version: 7.5.1 (same problem was there with 7.3.0)

**Steps to reproduce**

Set up gitlab authentication (tried with gitlab being the only method of auth).
Logout from mattermost.
Try to login

**Expected behavior**

I'm redirected to gitlab for authentication.

**Observed behavior**

Nothing happens when I click on the gitlab button (the button outline becomes blue, but I'm not redirected to gitlab)

**Possible fixes**

**Comment by dn0sar** [comment]
This is the log I get:
[mattermost.log](https://github.com/mattermost/desktop/files/10071622/mattermost.log)

Note that nothing gets added to the log when I try to click on the link. If I modify the HTML to have `target="_blank"` the link gets opened on my browser but then I get logged in the browser and not in the app.

**Comment by dn0sar** [comment]
Alright, I fixed the issue, but not really sure how.

The log you requested is actually under this folder in my system:
`/Users/<username>/Library/Containers/Mattermost.Desktop/Data/Library/Logs/Mattermost`

From this log I saw these messages when I click on the gitlab link:
```
[2022-11-23 13:59:07.560] [debug] webContentEvents.did-start-navigation {
  webContentsId: 5,
  url: 'https://<server_uri>/oauth/gitlab/login'
}
[2022-11-23 13:59:07.560] [debug] webContentEvents.will-navigate {
  webContentsId: 5,
  url: 'https://<server_uri>/oauth/gitlab/login'
}
[2022-11-23 13:59:07.560] [info]  Prevented desktop from navigating to: https://<server_uri>/oauth/gitlab/login
```

I then found issue #2069 based on that log. The Site URL is configured as "http://<server_uri>", a proxy takes care of the TLS connection. I then changed everything to be "https://<server_uri>" both the Site URL and the gitlab application configuration for the redirection. With this configuration I get a "Bad token" as a reply from gitlab (no one can login).

I then changed everything back to the old configuration (Site URL = 'http://<server_uri>', Gitalb redirect = 'http://<server_uri>/plugins/com.github.manland.mattermost-plugin-gitlab/oauth/complete'), restarted the mattermost server, and now everything works.

So essentially: I changed to another configuration -> Gitlab auth did not work, but I could click on the link
I changed back to the original configuration -> everything works


I'm very confused as to why this is working, but I'm glad it is

**Comment by Lukas-UAUX** [comment]
We also experience this 'bug' - a redirect *is* the expected behavior because it is a login with an external login provider (gitlab) which of course has another URL than the mattermost server. 

Looking at the linked code there is a value that is checked aigainst 'isCustomLoginURL'. Is this a new config value for Mattermost that needs to be configured now?

**Comment by Lukas-UAUX** [comment]
Not directly but we updated Gitlab & Mattermost server (Omnibus I think?) 

I talked to our server admins about this redirect blocking and 'isCustomLoginURL' - they will try out sth. if it can fix the issue today. For now we use the browser version if the client has no cached login...

**Comment by dn0sar** [comment]
I tried again today to login and logout.
I can't click the button anymore.

The  server configuration did not change since yesterday.

I think I was only able to login because something was cached on the desktop app while I made the configuration change on the server side.

**Comment by dn0sar** [comment]
Also, not sure if it helps, but I see this difference on the logs:

With Desktop app version 5.2.1 (where the I get the `Prevented desktop from navigating to ...`):
```
[2022-11-24 13:53:06.513] [debug] ServerInfo.trySendRemoteInfo <server_name> {
  name: '<server_name>',
  serverVersion: '7.5.1',
  siteURL: 'http://<server_url>',
  hasFocalboard: true,
  hasPlaybooks: true
}
```
With Desktop app version 5.1.1 (where the login works):
```
[2022-11-24 13:57:03.380] [debug] ServerInfo.trySendRemoteInfo <server_name> {
  name: '<server_name>',
  hasFocalboard: true,
  hasPlaybooks: true,
  serverVersion: '7.5.1'
}
```

**Comment by Lukas-UAUX** [comment]
> Please try a server restart if possible and let me know if that fixed it. I'll share your comment with our SRE team to check if it's related to gitlab/omnibus changes

Our server Team fixed it, as far as I understand they had to up the internal communication between mattermost and gitlab (same server) to https and use different ports for MM and gitlab now.

**Comment by dn0sar** [comment]
I set up a dev environment for the desktop app and figured out what is going on.

It seems that PR #2152 forces the SiteURL to be the one specified by the server instance. In my case, the server is behind a proxy, so it is configured as `http://`. However, the users are always redirected to the `https://` URL.

This is a problem because since version 5.2.1 users setting for SiteURL are discarded in favor of the server value (including the protocol part). Now because the URLs have a different protocol, [this line](https://github.com/mattermost/desktop/blob/911ae29623017d7820b04de6eacf4b9237be5fe9/src/main/views/webContentEvents.ts#L56) returns `undefined` and so all the subsequent checks are skipped, preventing the GitLab link to be opened.

It would be nice to allow users to specify at least a more secure protocol version than what the server indicates. I can open a PR with this change if you want. Seems that just changing [this function](https://github.com/mattermost/desktop/blob/911ae29623017d7820b04de6eacf4b9237be5fe9/src/main/app/utils.ts#L70) to select the most secure protocol is enough.

**Comment by devinbinnie** [comment]
> I set up a dev environment for the desktop app and figured out what is going on.
> 
> It seems that PR #2152 forces the SiteURL to be the one specified by the server instance. In my case, the server is behind a proxy, so it is configured as `http://`. However, the users are always redirected to the `https://` URL.
> 
> This is a problem because since version 5.2.1 the users' setting for SiteURL are discarded in favor of the server value (including the protocol part). Now because the URLs have a different protocol, [this line](https://github.com/mattermost/desktop/blob/911ae29623017d7820b04de6eacf4b9237be5fe9/src/main/views/webContentEvents.ts#L56) returns `undefined` and so all the subsequent checks are skipped, preventing the GitLab link to be opened.
> 
> It would be nice to allow users to specify at least a more secure protocol version than what the server indicates. I can open a PR with this change if you want. Seems that just changing [this function](https://github.com/mattermost/desktop/blob/911ae29623017d7820b04de6eacf4b9237be5fe9/src/main/app/utils.ts#L70) to select the most secure protocol is enough.

@dn0sar Is the server you're using always accessed via the proxy? If so, it might make more sense to just change the server's Site URL to `https://`, since that is a more secure protocol and should be the correct one.

Using a different Site URL than what is specified by the server causes a few other unexpected issues (eg. some plugins depend on it to generate links).

**Comment by teamfighter** [comment]
Subscribed to issue. Have almost same problem with Gitlab auth, nothing happens while clicking on button.

**Comment by SoftTransistor** [comment]
Same issue, thank you for looking at this :)

**Comment by martinantonmueller** [comment]
I believe this would be the right issue for us as well – event though we posted here:
https://github.com/mattermost/desktop/issues/2740#issuecomment-1614244099

I believe I posted the solution there as well and reproduce the text here:
The server is on the always on "latest", so it updates itself. Looking further into the issue it appears that the time-limit of the standalone-app is too short: https://challenges.cloudflare.com/ is used by gitlab to check for human activity. The main cause is the timeout set in the Mattermost Desktop App. When I copy the link from the Mattermost Desktop Client log and paste it into the browser, I finally get the Gitlab login page after 10 seconds. The Mattermost Desktop Client app only returns the error "handleSessionExpired false".

So is there a way to give the desktop app more time?

**Comment by devinbinnie** [comment]
Hey all,

I want to apologize for the lack of communication on these tickets regarding authentication. But I come with some good news: we've been working on separating the external login flow from the Desktop App, deferring to the browser like many other applications do. This should overall improve security and stability around the login process, as the current implementation requires a few hacks to make it work correctly. 

As of today, we released Mattermost v9.1 which has this feature implemented, and should work with all existing Desktop App versions going back to at least v5.4.0. Going forward we will be supporting this login flow for all external providers using the Desktop App, and this should fix any issues around login flow. Your server will need to be upgraded to take advantage of this feature.

I'll be closing these tickets for now as fixed, but feel free to comment and ask questions if you have any. Thanks for your patience :)

**Comment by comatose-tortoise** [comment]
I have this exact problem with version 5.7.0, is it really fixed? I'm using cloudflare tunnel where I need to enter the email before accessing the MM server with the desktop app. In the browser it works fine, but in the app on Mac, I have the exact same behavior as in this issue. The error in the log is the same to:

`[2024-04-03 16:27:59.299] [info]  [WebContentsEventM...] [Lab] [TAB_MESSAGING] Prevented desktop from navigating to: https://SERVER_URI.cloudflareaccess.com/cdn-cgi/access/verify-code/SERVER_URI?kid=d3dc6edf24..long line of jibberish`
